src/main/java/com/faforever/client/chat/AbstractChatTabController.java

```diff
@@ -574,20 +593,67 @@ private void appendMessage(ChatMessage chatMessage) {
       } else {
         cssClasses.add(MESSAGE_CSS_CLASS);
       }
-
       PlayerInfoBean playerInfo = playerService.getPlayerForUsername(chatMessage.getUsername());
-      if (playerInfo != null && playerInfo.isFriend()) {
-        cssClasses.add(FRIEND_CSS_CLASS);
-      }
+
+      assignPlayerColor(cssClasses, playerInfo);
 
       html = html.replace("{css-classes}", Joiner.on(' ').join(cssClasses));
 
       addToMessageContainer(html);
+
     } catch (IOException e) {
       throw new RuntimeException(e);
     }
   }
 
+  private void assignPlayerColor(Collection<String> cssClasses, PlayerInfoBean playerInfo) {
+    if (playerInfo == null) {
+      return;
+    }
+
+    //Mods, friends and foes are never randomly generated
+    String messageColor = "";
+    if (playerInfo.getModeratorInChannels() != null && playerInfo.getModeratorInChannels().size() > 0) {
+      //TODO this is here because there is no straight forward way to know what channel the message came from
+      for (Channel channel : chatService.getChannelsForUser(userService.getUsername())) {
+        if (chatService.getLevelsForChatUser(channel, playerInfo.getUsername()).size() > 0) {
+          messageColor = chatPrefs.getModsChatColor().toString();
+        }
+      }
+    } else if (playerInfo.isFriend()) {
+      messageColor = chatPrefs.getFriendsChatColor().toString();
+    } else if (playerInfo.isFoe()) {
+      messageColor = chatPrefs.getFoesChatColor().toString();
+    }
+
+    //
+    if (!chatPrefs.getPrettyColors() && messageColor.equals("")) {
+      if (playerInfo.isChatOnly()) {
+        messageColor = chatPrefs.getIrcChatColor().toString();
+      } else if (!playerInfo.getUsername().equals(userService.getUsername())) {
+        messageColor = chatPrefs.getOthersChatColor().toString();
+      }
+    }
+
+    if (messageColor.equals("")) {
+      if (playerInfo.getUsername().equals(userService.getUsername())) {
+        messageColor = chatPrefs.getSelfChatColor().toString();
+      } else {
+        ChatUser chatUser = chatService.getChatUser(playerInfo.getUsername());
+        /*if chatUser is null the message is lost, this happens when application is just initialized and the message is received before
+        the user is registered to chatService*/
+        if (chatUser != null) {
+          //FIXME chat user doesn't always have color even though it should
+          logger.debug("User {}", chatUser.getUsername());
+          messageColor = chatUser.getColor().toString();
+        }
+      }
+    }
+    if(messageColor != null) {
+      cssClasses.add("\" style=\"color:#" + messageColor.substring(2, 8));
```

Comment:
As before, `cssClasses` contains CSS style names. Putting HTML into it is very dirty.
`substring(2,8)` is not easy to understand, extract it into a method with a describing name or add a comment on top of it